tiledb/sm/cpp_api/query_condition.h

```diff
@@ -0,0 +1,225 @@
+/**
+ * @file   query_condition.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2021 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the C++ API for the TileDB QueryCondition object.
+ */
+
+#ifndef TILEDB_CPP_API_QUERY_CONDITION_H
+#define TILEDB_CPP_API_QUERY_CONDITION_H
+
+#include "context.h"
+#include "tiledb.h"
+
+#include <string>
+#include <type_traits>
+
+namespace tiledb {
+
+class QueryCondition {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /**
+   * Creates a TileDB query condition object.
+   *
+   * This is used to
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * tiledb::Context ctx;
+   * tiledb::Array array(ctx, "my_array", TILEDB_READ);
+   * tiledb::Query query(ctx, array, TILEDB_READ);
+   *
+   * int cmp_value = 5;
+   * tiledb::QueryCondition qc(ctx, "a1", &cmp_value, sizeof(int), TILEDB_LT);
+   * query.set_condition(qc);
+   * @endcode
+   *
+   * @param ctx TileDB context.
+   * @param attribute_name The name of the attribute to compare against.
+   * @param condition_value The fixed value to compare against.
+   * @param condition_value_size The byte size of `condition_value`.
+   * @param op The comparison operation between each cell value and
+   * `condition_value`.
+   */
+  QueryCondition(
```

Comment:
Can we add a constructor which takes an `std::string` as the condition value? We have overrides in ranges and other places to handle `std::string` for the user.

Reply:
Done